File: framework/src/play-ws/src/main/java/play/libs/ws/WSClient.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws;

import java.io.IOException;

/**
 * This is the WS Client interface for Java.
 *
 * Most of the time you will access this through dependency injection, i.e.
 *
 * <pre>
 * <code>import javax.inject.Inject;
 * import play.libs.ws.*;
 * import java.util.concurrent.CompletionStage;
 *
 * public class MyService {
 *   {@literal @}Inject WSClient ws;
 *
 *    // ...
 * }
 * </code>
 * </pre>
 *
 * Please see https://www.playframework.com/documentation/latest/JavaWS for more details.
 */
public interface WSClient extends java.io.Closeable {

    /**
     * The underlying implementation of the client, if any.  You must cast the returned value to the type you want.
     *
     * @return the backing object.
     */
    Object getUnderlying();

    /**
     * @return the Scala version for this WSClient.
     */
    play.api.libs.ws.WSClient asScala();

    /**
     * Returns a WSRequest object representing the URL.  You can append additional
     * properties on the WSRequest by chaining calls, and execute the request to
     * return an asynchronous {@code Promise<WSResponse>}.
     *
     * @param url the URL to request
     * @return the request
     */
    WSRequest url(String url);

    /**
     * Closes this client, and releases underlying resources.
     * <p>
     * Use this for manually instantiated clients.
     */
    void close() throws IOException;
}

File: framework/src/play-ws/src/main/scala/play/api/libs/ws/WSClient.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws

import java.io.{ Closeable, IOException }

/**
 * A Play specific WS client that can use Play specific classes in the request and response building.
 *
 * Typically, access this class through dependency injection, i.e.
 *
 * {{{
 *   class MyService @Inject()(ws: WSClient) {
 *     val response: Future[WSResponse] = ws.url("http://example.com").get()
 *   }
 * }}}
 *
 * Please see the documentation at https://www.playframework.com/documentation/latest/ScalaWS for more details.
 */
trait WSClient extends Closeable {

  /**
   * The underlying implementation of the client, if any.  You must cast explicitly to the type you want.
   * @tparam T the type you are expecting (i.e. isInstanceOf)
   * @return the backing class.
   */
  def underlying[T]: T

  /**
   * Generates a request.
   *
   * @param url The base URL to make HTTP requests to.
   * @return a request
   */
  def url(url: String): WSRequest

  /** Closes this client, and releases underlying resources. */
  @throws[IOException] def close(): Unit
}
